commit 913c71bc40d1894196eac469a00b480f5925fe73
Author: 0x1eef <0x1eef@protonmail.com>
Date:   Wed Apr 2 22:04:17 2025 +0000

    nvidia: apply patch
    
    This patch is taken from a fork of FreeBSD that is focused on adding
    PCI passthrough support for bhyve.

diff --git a/sys/dev/pci/pcireg.h b/sys/dev/pci/pcireg.h
index f6aaf30611e4..71f307eaa7db 100644
--- a/sys/dev/pci/pcireg.h
+++ b/sys/dev/pci/pcireg.h
@@ -1181,3 +1181,8 @@
 #define	PCIM_AMDIOMMU_MISC1_MSINUMGA_MASK 0x1f	
 
 #endif	/* __PCI_PCIREG_H */
+
+/*
+ * PCI Vendors
+ */
+#define PCI_VENDOR_NVIDIA 0x10DE
diff --git a/usr.sbin/bhyve/pci_passthru.c b/usr.sbin/bhyve/pci_passthru.c
index 61983010192a..219e12ff2915 100644
--- a/usr.sbin/bhyve/pci_passthru.c
+++ b/usr.sbin/bhyve/pci_passthru.c
@@ -75,6 +75,7 @@
 #define PASSTHRU_MMIO_MAX 2
 
 static int pcifd = -1;
+static uint8_t *nvidia_bar0;
 
 SET_DECLARE(passthru_dev_set, struct passthru_dev);
 
@@ -619,6 +620,30 @@ cfginitbar(struct passthru_softc *sc)
 		sc->psc_bar[i].addr = base;
 		sc->psc_bar[i].lobits = 0;
 
+		if (i == 0 &&
+		    pci_host_read_config(&sc->psc_sel, PCIR_VENDOR, 2) ==
+			PCI_VENDOR_NVIDIA &&
+		    pci_host_read_config(&sc->psc_sel, PCIR_CLASS, 2) == PCIC_DISPLAY) {
+
+			struct pci_bar_mmap pbm;
+			memset(&pbm, 0, sizeof(pbm));
+			pbm.pbm_sel = sc->psc_sel;
+			pbm.pbm_flags = PCIIO_BAR_MMAP_RW;
+			pbm.pbm_reg = PCIR_BAR(i);
+			pbm.pbm_memattr = VM_MEMATTR_DEVICE;
+			if (ioctl(pcifd, PCIOCBARMMAP, &pbm) != 0) {
+				warn("Failed to map Nvidia BAR 0");
+				return (-1);
+			}
+			assert(pbm.pbm_bar_off == 0);
+
+			nvidia_bar0 = (uint8_t *)(uintptr_t)pbm.pbm_map_base;
+			if (pbm.pbm_map_length < 0x88000) {
+				warnx("Invalid BAR 0 size for Nvidia device");
+				return (-1);
+			}
+		}
+
 		/* Allocate the BAR in the guest I/O or MMIO space */
 		error = pci_emul_alloc_bar(pi, i, bartype, size);
 		if (error)
@@ -964,6 +989,11 @@ passthru_init(struct pci_devinst *pi, nvlist_t *nvl)
 	if ((error = set_pcir_handler(sc, PCIR_COMMAND, 0x04, NULL, NULL)) != 0)
 		goto done;
 
+	/* copy PCI header to virtual PCI space */
+	for (uint32_t i = 0; i < 0x3E; ++i) {
+		pci_set_cfgdata8(pi, i, pci_host_read_config(&sc->psc_sel, i, 1));
+	}
+
 	SET_FOREACH(devpp, passthru_dev_set) {
 		devp = *devpp;
 		assert(devp->probe != NULL);
@@ -1062,6 +1092,13 @@ passthru_cfgread(struct pci_devinst *pi, int coff, int bytes, uint32_t *rv)
 
 	sc = pi->pi_arg;
 
+	if (coff >= PCI_REGMAX) {
+		warnx("%s (%d/%d/%d): Invalid offset %x", __func__, pi->pi_bus,
+		    pi->pi_slot, pi->pi_func, coff);
+		*rv = 0xFFFFFFFF;
+		return (-1);
+	}
+
 	if (sc->psc_pcir_rhandler[coff] != NULL)
 		return (sc->psc_pcir_rhandler[coff](sc, pi, coff, bytes, rv));
 
@@ -1154,6 +1191,12 @@ passthru_cfgwrite(struct pci_devinst *pi, int coff, int bytes, uint32_t val)
 
 	sc = pi->pi_arg;
 
+	if (coff >= PCI_REGMAX) {
+		warnx("%s (%d/%d/%d): Invalid offset %x", __func__, pi->pi_bus,
+		    pi->pi_slot, pi->pi_func, coff);
+		return (-1);
+	}
+
 	if (sc->psc_pcir_whandler[coff] != NULL)
 		return (sc->psc_pcir_whandler[coff](sc, pi, coff, bytes, val));
 
@@ -1171,6 +1214,10 @@ passthru_write(struct pci_devinst *pi, int baridx, uint64_t offset, int size,
 
 	if (baridx == pci_msix_table_bar(pi)) {
 		msix_table_write(sc, offset, size, value);
+	} else if (baridx == 0 &&
+	    pci_get_cfgdata16(pi, PCIR_VENDOR) == PCI_VENDOR_NVIDIA &&
+	    pci_get_cfgdata8(pi, PCIR_CLASS) == PCIC_DISPLAY) {
+		passthru_cfgwrite(pi, offset - 0x88000, size, value);
 	} else {
 		assert(pi->pi_bar[baridx].type == PCIBAR_IO);
 		assert(size == 1 || size == 2 || size == 4);
@@ -1199,6 +1246,13 @@ passthru_read(struct pci_devinst *pi, int baridx, uint64_t offset, int size)
 
 	if (baridx == pci_msix_table_bar(pi)) {
 		val = msix_table_read(sc, offset, size);
+	} else if (baridx == 0 &&
+	    pci_get_cfgdata16(pi, PCIR_VENDOR) == PCI_VENDOR_NVIDIA &&
+	    pci_get_cfgdata8(pi, PCIR_CLASS) == PCIC_DISPLAY) {
+		/* dummy read to MMIO because hw might depend on it */
+		memcpy(&val, nvidia_bar0 + offset, size);
+
+		passthru_cfgread(pi, offset - 0x88000, size, (uint32_t *)&val);
 	} else {
 		assert(pi->pi_bar[baridx].type == PCIBAR_IO);
 		assert(size == 1 || size == 2 || size == 4);
@@ -1278,6 +1332,52 @@ passthru_mmio_addr(struct pci_devinst *pi, int baridx, int enabled,
 	struct passthru_softc *sc;
 
 	sc = pi->pi_arg;
+
+	if (pci_get_cfgdata16(pi, PCIR_VENDOR) == PCI_VENDOR_NVIDIA &&
+	    pci_get_cfgdata8(pi, PCIR_CLASS) == PCIC_DISPLAY && baridx == 0) {
+		uint64_t gpa = address;
+		uint64_t len = 0x880000;
+		uint64_t hpa = sc->psc_bar[baridx].addr;
+
+		if (!enabled) {
+			if (vm_unmap_pptdev_mmio(pi->pi_vmctx, sc->psc_sel.pc_bus,
+				sc->psc_sel.pc_dev, sc->psc_sel.pc_func, gpa,
+				len) != 0) {
+				warnx(
+				    "pci_passthru: vm_unmap_pptdev_mmio nvidia low failed");
+			}
+		} else {
+			if (vm_map_pptdev_mmio(pi->pi_vmctx, sc->psc_sel.pc_bus,
+				sc->psc_sel.pc_dev, sc->psc_sel.pc_func, gpa,
+				len, hpa) != 0) {
+				warnx(
+				    "pci_passthru: vm_map_pptdev_mmio nvidia low failed");
+			}
+		}
+
+		gpa += 0x880000 + 0x1000;
+		hpa += 0x880000 + 0x1000;
+		len = sc->psc_bar[baridx].size - (0x880000 + 0x1000);
+
+		if (!enabled) {
+			if (vm_unmap_pptdev_mmio(pi->pi_vmctx, sc->psc_sel.pc_bus,
+				sc->psc_sel.pc_dev, sc->psc_sel.pc_func, gpa,
+				len) != 0) {
+				warnx(
+				    "pci_passthru: vm_unmap_pptdev_mmio nvidia low failed");
+			}
+		} else {
+			if (vm_map_pptdev_mmio(pi->pi_vmctx, sc->psc_sel.pc_bus,
+				sc->psc_sel.pc_dev, sc->psc_sel.pc_func, gpa,
+				len, hpa) != 0) {
+				warnx(
+				    "pci_passthru: vm_map_pptdev_mmio nvidia low failed");
+			}
+		}
+
+		return;
+	}
+
 	if (!enabled) {
 		if (vm_unmap_pptdev_mmio(pi->pi_vmctx, sc->psc_sel.pc_bus,
 					 sc->psc_sel.pc_dev,
